internal/resources/google/sql_database_instance.go

```diff
@@ -0,0 +1,316 @@
+}
+
+var SQLDatabaseInstanceUsageSchema = []*schema.UsageItem{{Key: "backup_storage_gb", ValueType: schema.Float64, DefaultValue: 0}}
+
+func (r *SQLDatabaseInstance) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage(r, u)
+}
+
+func (r *SQLDatabaseInstance) BuildResource() *schema.Resource {
+	var resource *schema.Resource
+
+	replica := false
+	if r.ReplicaConfiguration != "" {
+		replica = true
+	}
+
+	resource = sqlDatabaseInstanceCostComponents(r, false, r.Address)
+	if replica {
+		resource.SubResources = append(resource.SubResources, sqlDatabaseInstanceCostComponents(r, true, "Replica"))
+	}
+
+	return resource
+}
+
+type SQLInstanceDBType int
+
+const (
+	MySQL SQLInstanceDBType = iota
+	PostgreSQL
+	SQLServer
+)
+
+func sqlDatabaseInstanceCostComponents(r *SQLDatabaseInstance, replica bool, name string) *schema.Resource {
+	var costComponents []*schema.CostComponent
+	tier := r.Tier
+
+	availabilityType := "ZONAL"
+	if r.AvailabilityType != "" && !replica {
+		availabilityType = r.AvailabilityType
+	}
+
+	region := r.Region
+	dbVersion := r.DatabaseVersion
+	dbType := sqlInstanceDBVersionToDBType(dbVersion)
+
+	diskType := "PD_SSD"
+	if r.DiskType != "" {
+		diskType = r.DiskType
+	}
+
+	var diskSizeGB int64 = 10
+	if r.DiskSize != 0 {
+		diskSizeGB = r.DiskSize
+	}
+
+	if sqlInstanceTierToResourceGroup(tier) != "" && dbType != SQLServer {
+		costComponents = append(costComponents, sharedSQLInstance(tier, availabilityType, dbType, region))
+	} else if sqlInstanceTierToResourceGroup(tier) == "" && strings.Contains(tier, "db-custom-") {
+		cpu, _ := strconv.ParseInt(strings.Split(tier, "-")[2], 10, 32)
+		vCPU := decimalPtr(decimal.NewFromInt32(int32(cpu)))
+
+		costComponents = append(costComponents, cpuCostComponent(region, tier, availabilityType, dbType, vCPU))
+
+		ram, _ := strconv.ParseInt(strings.Split(tier, "-")[3], 10, 32)
+		memory := decimalPtr(decimal.NewFromInt32(int32(ram)).Div(decimal.NewFromInt(1024)))
+
+		costComponents = append(costComponents, memoryCostComponent(region, tier, availabilityType, dbType, memory))
+	} else if strings.Contains(tier, "db-n1-") && dbType == MySQL {
+		costComponents = append(costComponents, sharedSQLInstance(tier, availabilityType, dbType, region))
+	}
+
+	costComponents = append(costComponents, sqlInstanceStorage(region, dbType, availabilityType, diskType, diskSizeGB))
+
+	if !replica {
+		var backupGB *decimal.Decimal
+		if r.BackupStorageGB != nil {
+			backupGB = decimalPtr(decimal.NewFromFloat(*r.BackupStorageGB))
+		}
+		costComponents = append(costComponents, backupCostComponent(region, backupGB))
+
+		if r.UseIPV4 {
+			costComponents = append(costComponents, ipv4CostComponent())
+		}
+	}
+
+	return &schema.Resource{
+		Name:           name,
+		CostComponents: costComponents,
+	}
+}
+
+func memoryCostComponent(region string, tier string, availabilityType string, dbType SQLInstanceDBType, memory *decimal.Decimal) *schema.CostComponent {
```

Comment:
Can all the below functions be on the struct so there's no risk of conflict on the package level? If they can then I think we can tidy up the arguments that are passed to it, since most of them soon be available as struct properties.

Reply:
Great idea, applied and removed the redundant params.